sandbox/cgroup/cgroup_test.go

```diff
@@ -236,46 +238,157 @@ func (s *cgroupSuite) TestProgGroupBadSelector(c *C) {
+	c.Check(pids, HasLen, 0)
+}
 
-	pids, err = cgroup.PidsInGroup(s.rootDir, "group.does.not.exist")
+func (s *cgroupSuite) TestPidsOfSnapUnrelatedStuff(c *C) {
+	// Things that are not related to the snap are not being picked up.
+	s.writePids(c, "udisks2.service", []int{100})
+	s.writePids(c, "snap..service", []int{101})
+	s.writePids(c, "snap..scope", []int{102})
+	s.writePids(c, "snap.*.service", []int{103})
+	s.writePids(c, "snap.*.scope", []int{104})
+	s.writePids(c, "snapd.service", []int{105})
+	s.writePids(c, "snap-spotify-35.service", []int{106})
+
+	pids, err := cgroup.PidsOfSnap("pkg")
 	c.Assert(err, IsNil)
-	c.Assert(pids, IsNil)
+	c.Check(pids, HasLen, 0)
 }
 
-func (s *cgroupSuite) TestPidsBadInput(c *C) {
-	err := os.MkdirAll(filepath.Join(s.rootDir, "group1"), 0755)
+func (s *cgroupSuite) TestPidsOfSnapSecurityTags(c *C) {
+	// Pids are collected and assigned to bins by security tag
+	s.writePids(c, "system.slice/snap.$RANDOM.pkg.hook.configure.scope", []int{1})
+	s.writePids(c, "system.slice/snap.pkg.daemon.service", []int{2})
+
+	pids, err := cgroup.PidsOfSnap("pkg")
 	c.Assert(err, IsNil)
-	gPids := []byte(`123
-zebra
-567
-`)
-	err = ioutil.WriteFile(filepath.Join(s.rootDir, "group1/cgroup.procs"), gPids, 0755)
+	c.Check(pids, DeepEquals, map[string][]int{
+		"snap.pkg.hook.configure": {1},
+		"snap.pkg.daemon":         {2},
+	})
+}
+
+func (s *cgroupSuite) TestPidsOfInstances(c *C) {
+	// Instances are not confused between themselves and between the non-instance version.
+	s.writePids(c, "system.slice/snap.pkg_prod.daemon.service", []int{1})
+	s.writePids(c, "system.slice/snap.pkg_devel.daemon.service", []int{2})
+	s.writePids(c, "system.slice/snap.pkg.daemon.service", []int{3})
+
+	// The main one
+	pids, err := cgroup.PidsOfSnap("pkg")
 	c.Assert(err, IsNil)
+	c.Check(pids, DeepEquals, map[string][]int{
+		"snap.pkg.daemon": {3},
+	})
 
-	pids, err := cgroup.PidsInGroup(s.rootDir, "group1")
-	c.Assert(err, ErrorMatches, `cannot parse pid "zebra"`)
-	c.Assert(pids, IsNil)
+	// The development one
+	pids, err = cgroup.PidsOfSnap("pkg_devel")
+	c.Assert(err, IsNil)
+	c.Check(pids, DeepEquals, map[string][]int{
+		"snap.pkg_devel.daemon": {2},
+	})
+
+	// The production one
+	pids, err = cgroup.PidsOfSnap("pkg_prod")
+	c.Assert(err, IsNil)
+	c.Check(pids, DeepEquals, map[string][]int{
+		"snap.pkg_prod.daemon": {1},
+	})
+}
+
+func (s *cgroupSuite) TestPidsOfAggregation(c *C) {
+	// A single snap may be invoked by multiple users in different sessions.
+	// All of their PIDs are collected though.
+	s.writePids(c, "user.slice/user-1000.slice/user@1000.service/gnome-shell-wayland.service/snap.$RANDOM1.pkg.app.scope", []int{1})
+	s.writePids(c, "user.slice/user-1001.slice/user@1001.service/gnome-shell-wayland.service/snap.$RANDOM2.pkg.app.scope", []int{2})
+
+	pids, err := cgroup.PidsOfSnap("pkg")
+	c.Assert(err, IsNil)
+	c.Check(pids, DeepEquals, map[string][]int{
+		"snap.pkg.app": {1, 2},
+	})
```

Comment:
This isn't quite right, I think you made a small mistake here. 

```
user.slice/user-1000.slice/user@1000.service/gnome-shell-wayland.service/snap.$RANDOM1.pkg.app.scope", []int{1}
```

This has the exact same path as

```
s.writePids(c, "user.slice/user-1000.slice/user@1000.service/gnome-shell-wayland.service/snap.$RANDOM1.pkg.app.scope", []int{2})  // mock fork() by pid 1
```

...and due to how `writePids` behaves (it overwrites existing values) the test adjustment would only measure pids `2` and `4`.

I've adjusted this to use distinct four random numbers, which is also more realistic, to avoid the problem and produce the effect you, I think, were after.

Reply:
This will be patch 6c7b16aca471bf7706c6294ef7cf78e9064a16cd